test/jdk/javax/swing/text/html/HtmlFontSizeTest.java

```diff
@@ -0,0 +1,101 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+/*
+ * @test
+ * @bug 8231286
+ * @key headful
+ * @summary  Verifies if HTML font size too large with high-DPI scaling and W3C_LENGTH_UNITS
+ * @run main/othervm -Dsun.java2d.uiScale=1.0 HtmlFontSizeTest
+ */
+
+import java.awt.BorderLayout;
+import java.awt.Rectangle;
+
+import javax.swing.JFrame;
+import javax.swing.JEditorPane;
+import javax.swing.SwingUtilities;
+import javax.swing.text.Document;
+import javax.swing.text.html.HTMLEditorKit;
+import javax.swing.text.html.StyleSheet;
+
+public class HtmlFontSizeTest {
+    static volatile Rectangle w3cFrameSize;
+    static volatile Rectangle stdFrameSize;
+
+    private static Rectangle test(boolean w3ccheck) {
+        JFrame frame = new JFrame();
+        try {
+            frame.setLayout(new BorderLayout());
+
+            JEditorPane htmlPane = new JEditorPane();
+            htmlPane.setEditable(false);
+
+            if (w3ccheck) {
+                htmlPane.putClientProperty(JEditorPane.W3C_LENGTH_UNITS, Boolean.TRUE);
+            }
+
+            HTMLEditorKit kit = new HTMLEditorKit();
+            htmlPane.setEditorKit(kit);
+
+            StyleSheet styleSheet = kit.getStyleSheet();
+            styleSheet.addRule("body { font-family: SansSerif; font-size: 16pt; }");
+
+            String htmlString = "<html>\n"
+                + "<body>\n"
+                + "<p>This should be 16 pt.</p>\n"
+                + "</body>\n"
+                + "</html>";
+
+            Document doc = kit.createDefaultDocument();
+            htmlPane.setDocument(doc);
+            htmlPane.setText(htmlString);
+
+            frame.add(htmlPane, BorderLayout.CENTER);
+            frame.setUndecorated(true);
+            frame.pack();
+            frame.setVisible(true);
+            frame.setLocationRelativeTo(null);
+
+            return frame.getBounds();
+        } finally {
+            frame.dispose();
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        SwingUtilities.invokeAndWait(() -> {
+            w3cFrameSize = test(true);
+            stdFrameSize = test(false);
+        });
+        System.out.println("frame height with W3C:" + w3cFrameSize);
+        System.out.println("frame height without W3C:" + stdFrameSize);
+
+        float ratio = (float)w3cFrameSize.width / (float)stdFrameSize.width;
+        System.out.println("w3cFrameSize.width/stdFrameSize.width " + ratio);
+
+        String str = String.format("%.1g%n", ratio);
+        if (str.compareTo(String.format("%.1g%n", 1.3f)) != 0) {
+            throw new RuntimeException("HTML font size too large with high-DPI scaling and W3C_LENGTH_UNITS");
```

Comment:
The test passes for me with the fresh build of JDK, even without the fix applied.
Indeed, I get the same frame dimensions in both cases: 150×42.

However, the test fails when run with JDK 11.0.10 or JDK 15, the dimension of `htmlPane` is 409×76.

The formatted string, `str`, contains only the number before the decimal point. I mean for `1.3f`, the formatted string is `"1"`. I'd expect the formatted string to have at least one digit after the decimal point but there's none.

Does
```
if ("1.3".equals(String.format("%1.1", ratio))
```
look clearer?


I guess the test does not fail if the system scaling (of the main monitor) is set to 100% because `res` value read from `Toolkit.getDefaultToolkit().getScreenResolution()` is 96/72 which is 1.3.

In my testing, `sun.java2d.uiScale` has no effect on the result.

Reply:
I propose to move the test into CSS folder, and probably to create its own folder, `8231286` or 'font-size`.

Reply:
> The test passes for me with the fresh build of JDK, even without the fix applied.
> Indeed, I get the same frame dimensions in both cases: 150×42.
> 
> However, the test fails when run with JDK 11.0.10 or JDK 15, the dimension of `htmlPane` is 409×76.
> 
[JDK-8257664](https://bugs.openjdk.java.net/browse/JDK-8257664): HTMLEditorKit: Wrong CSS relative font sizes
has changed the behaviour. With the fix reverted, I get frame size of 409×76 as in previous versions.

Reply:
> > The test passes for me with the fresh build of JDK, even without the fix applied.
> > Indeed, I get the same frame dimensions in both cases: 150×42.
> > However, the test fails when run with JDK 11.0.10 or JDK 15, the dimension of `htmlPane` is 409×76.
> 
> [JDK-8257664](https://bugs.openjdk.java.net/browse/JDK-8257664): HTMLEditorKit: Wrong CSS relative font sizes
> has changed the behaviour. With the fix reverted, I get frame size of 409×76 as in previous versions.

I have submitted [JDK-8260687](https://bugs.openjdk.java.net/browse/JDK-8260687): _Inherited font size is smaller than expected when using StyleSheet to add styles_.

I suggest declaring the `font-size` property on the `<body>` element in `style` attribute.

Reply:
> Does
> ```java
> if ("1.3".equals(String.format("%1.1", ratio))
> ```
> look clearer?

I went with something similar to this now. Unfortunately, I also needed to use `Locale.ENGLISH` explicitely to get a dot a as the decimal separator.

> I guess the test does not fail if the system scaling (of the main monitor) is set to 100% because res value read from Toolkit.getDefaultToolkit().getScreenResolution() is 96/72 which is 1.3.
> 
> In my testing, sun.java2d.uiScale has no effect on the result.

To be honest, this is unclear to me as well, I just copied what @prsadhuk had. When testing with older JDKs, I got some different dimensions, but at least now with the fix applied it doesn't seem to change anything.

> I propose to move the test into CSS folder, and probably to create its own folder, `8231286` or 'font-size`.

Done, moved into `./CSS/8231286`. 

> I suggest declaring the `font-size` property on the `<body>` element in `style` attribute.

Done.

Reply:
> > Does
> > ```java
> > if ("1.3".equals(String.format("%1.1", ratio))
> > ```
> > 
> > 
> > look clearer?
> 
> I went with something similar to this now. Unfortunately, I also needed to use `Locale.ENGLISH` explicitely to get a dot as the decimal separator.

Interesting. With `Locale.ENGLISH`, it will be more robust.


> > I guess the test does not fail if the system scaling (of the main monitor) is set to 100% because res value read from Toolkit.getDefaultToolkit().getScreenResolution() is 96/72 which is 1.3.
> > In my testing, sun.java2d.uiScale has no effect on the result.
> 
> To be honest, this is unclear to me as well, I just copied what @prsadhuk had. When testing with older JDKs, I got some different dimensions, but at least now with the fix applied it doesn't seem to change anything.

It is likely because `sun.java2d.uiScale` gets applied to both frames. If the font size is the same, 2d surface gets scaled up to the same level, which results in the same size of the frames.

Yet the calculations of the conversion ratios in CSS depend only on the resolution of the main screen. If the main screen resolution is 96dpi (100% scaling), the font size will be the same as after your fix.

Reply:
> It is likely because sun.java2d.uiScale gets applied to both frames. If the font size is the same, 2d surface gets scaled up to the same level, which results in the same size of the frames.
> 
> Yet the calculations of the conversion ratios in CSS depend only on the resolution of the main screen. If the main screen resolution is 96dpi (100% scaling), the font size will be the same as after your fix.

That makes sense.

What should we do about it? Is there a way to inject a different screen resolution for testing?
And should I remove `sun.java2d.uiScale` from the doc?

Reply:
> What should we do about it? Is there a way to inject a different screen resolution for testing?

I do not know. You can change the scale factor on your main display, the `Toolkit.getScreenResolution()` will return the new value.

> And should I remove `sun.java2d.uiScale` from the doc?

I think so. It does not affect the test in any way.